### DISCUSSION TITLE
ST7701s ESP32 , colors are not displayed correctly

### DISCUSSION THREAD

**Discussion by T-Bone-one** [moononournation/Arduino_GFX#535] [General]
![edbe5ea6-487b-48c8-aa52-d45fdd9b8f6e __CR0,0,1464,600_PT0_SX1464_V1___](https://github.com/user-attachments/assets/b1a16973-3201-4437-8527-940306c6024b)

Hi, can someone help me with this setup.. Skecth is working but RED Color is missing ... i see only blue and green colors. (display is a round 2.8inch).

`
#include <Arduino_GFX_Library.h>
#define GFX_BL -1
Arduino_DataBus *bus = new Arduino_SWSPI(
    GFX_NOT_DEFINED /* DC */, 42 /* CS */,
    2 /* SCK */, 1 /* MOSI */, GFX_NOT_DEFINED /* MISO */);

/* More dev device declaration: https://github.com/moononournation/Arduino_GFX/wiki/Dev-Device-Declaration */
#if defined(DISPLAY_DEV_KIT)
Arduino_GFX *gfx = create_default_Arduino_GFX();
#else /* !defined(DISPLAY_DEV_KIT) */

Arduino_ESP32RGBPanel *rgbpanel = new Arduino_ESP32RGBPanel(
    40 /* DE */, 39 /* VSYNC */, 38 /* HSYNC */, 41 /* PCLK */,
    46 /* R0 */, 3 /* R1 */, 8 /* R2 */, 18 /* R3 */, 17 /* R4 */,
    14 /* G0 */, 13 /* G1 */, 12 /* G2 */, 11 /* G3 */, 10 /* G4 */, 9 /* G5 */,
    5 /* B0 */, 45 /* B1 */, 48 /* B2 */, 47 /* B3 */, 21 /* B4 */,
    1 /* hsync_polarity */, 22 /* hsync_front_porch */, 3 /* hsync_pulse_width */, 45 /* hsync_back_porch */,
    1 /* vsync_polarity */, 4 /* vsync_front_porch */, 12 /* vsync_pulse_width */, 40 /* vsync_back_porch */);
Arduino_RGB_Display *gfx = new Arduino_RGB_Display(
    480 /* width */, 480 /* height */, rgbpanel, 0 /* rotation */, true /* auto_flush */,
    bus, GFX_NOT_DEFINED /* RST */, st7701_type6_init_operations, sizeof(st7701_type6_init_operations));

#endif /* !defined(DISPLAY_DEV_KIT) */
/*******************************************************************************
 * End of Arduino_GFX setting
 ******************************************************************************/

void setup(void)
{
  Serial.begin(115200);
  // Serial.setDebugOutput(true);
  // while(!Serial);
  Serial.println("Arduino_GFX Hello World example");

#ifdef GFX_EXTRA_PRE_INIT
  GFX_EXTRA_PRE_INIT();
#endif

  // Init Display
  if (!gfx->begin())
  {
    Serial.println("gfx->begin() failed!");
  }
  gfx->fillScreen(RED);

#ifdef GFX_BL
  pinMode(GFX_BL, OUTPUT);
  digitalWrite(GFX_BL, HIGH);
#endif

  gfx->setCursor(10, 10);
  gfx->setTextColor(RED);
  gfx->println("Hello World!");

  delay(5000); // 5 seconds
}

void loop()
{
  gfx->setCursor(random(gfx->width()), random(gfx->height()));
  gfx->setTextColor(random(0xffff), random(0xffff));
  gfx->setTextSize(random(6) /* x scale */, random(6) /* y scale */, random(2) /* pixel_margin */);
  gfx->println("Hello World!");

  delay(1000); // 1 second
}`

**Comment by T-Bone-one** [comment]
![20240911_094526](https://github.com/user-attachments/assets/161557ff-d3b6-46fc-8826-4c1c0aa30225)

Can someone help me please ? What am i doing wrong ? See the result in photo.
Colors are not displayed correctly.

Thank you.

**Comment by T-Bone-one** [comment]
`
#include <Arduino_GFX_Library.h>
#define GFX_BL -1  // default backlight pin, you may replace DF_GFX_BL to actual backlight pin
Arduino_DataBus *bus = new Arduino_SWSPI(GFX_NOT_DEFINED, 42, 2, 1, GFX_NOT_DEFINED);
//Arduino_DataBus *bus = new Arduino_SWSPI(8 /* DC */, 9 /* CS */, 13 /* SCK */, 11 /* MOSI */, GFX_NOT_DEFINED /* MISO */);
Arduino_ESP32RGBPanel *rgbpanel = new Arduino_ESP32RGBPanel(
  40 /* DE */, 39 /* VSYNC */, 38 /* HSYNC */, 41 /* PCLK */,
  46 /* R0 */, 3 /* R1 */, 8 /* R2 */, 18 /* R3 */, 17 /* R4 */,
  14 /* G0 */, 13 /* G1 */, 12 /* G2 */, 11 /* G3 */, 10 /* G4 */, 9 /* G5 */,
  5 /* B0 */, 45 /* B1 */, 48 /* B2 */, 47 /* B3 */, 21 /* B4 */,
  1 /* hsync_polarity */, 22 /* hsync_front_porch */, 24 /* hsync_pulse_width */, 6 /* hsync_back_porch */,
  1 /* vsync_polarity */, 16 /* vsync_front_porch */, 4 /* vsync_pulse_width */, 10 /* vsync_back_porch */);


Arduino_RGB_Display *gfx = new Arduino_RGB_Display(
  480 /* width */, 480 /* height */, rgbpanel, 0 /* rotation */, true /* auto_flush */,
  bus, GFX_NOT_DEFINED /* RST */, st7701_type6_init_operations, sizeof(st7701_type6_init_operations));


/* !defined(DISPLAY_DEV_KIT) */
/*******************************************************************************
 * End of Arduino_GFX setting
 ******************************************************************************/

void setup(void) {
  Serial.begin(115200);
  // Serial.setDebugOutput(true);
  // while(!Serial);
  Serial.println("Arduino_GFX Hello World example");

#ifdef GFX_EXTRA_PRE_INIT
  GFX_EXTRA_PRE_INIT();
#endif

  // Init Display
  if (!gfx->begin()) {
    Serial.println("gfx->begin() failed!");
  }
  gfx->fillScreen(BLACK);

#ifdef GFX_BL
  pinMode(GFX_BL, OUTPUT);
  digitalWrite(GFX_BL, HIGH);
#endif

  gfx->setCursor(150, 150);
  gfx->setTextSize(2);
  gfx->setTextColor(RED);
  gfx->println("Hello World! in RED");
  gfx->setCursor(150, 180);
  gfx->setTextColor(BLUE);
  gfx->println("Hello World! in BLUE");
  gfx->setCursor(150, 210);
  gfx->setTextColor(GREEN);
  gfx->println("Hello World! in GREEN");
  gfx->setCursor(150, 240);
  gfx->setTextColor(YELLOW);
  gfx->println("Hello World! in YELLOW");



  //delay(5000);  // 5 seconds
}

void loop() {
  // gfx->setCursor(random(gfx->width()), random(gfx->height()));
  // gfx->setTextColor(random(0xffff), random(0xffff));
  // gfx->setTextSize(random(6) /* x scale */, random(6) /* y scale */, random(2) /* pixel_margin */);
  // gfx->println("Hello World!");

  // delay(1000); // 1 second
}`

**Comment by jfseaman** [comment]
I have noticed that "GREEN" with that display is not very green.

**Comment by jfseaman** [comment]
I have the square display that works with the installed software.  ST7701_type1 in GFX

I saved a copy as a bin file.

I can't get my other shetches to work yet.

**Reply by T-Bone-one** [reply]
After weeks of testing i finally managed to get it working as it should.

**Reply by jfseaman** [reply]
> After weeks of testing i finally managed to get it working as it should.

Is that last listing your settings?

**Reply by T-Bone-one** [reply]
`static const uint8_t st7701_type6_init_operations[] = {
BEGIN_WRITE,

    WRITE_COMMAND_8, 0xFF,
    WRITE_BYTES, 5, 0x77, 0x01, 0x00, 0x00, 0x13,

    WRITE_C8_D8, 0xEF, 0x08,

    WRITE_COMMAND_8, 0xFF,
    WRITE_BYTES, 5, 0x77, 0x01, 0x00, 0x00, 0x10,

    WRITE_C8_D16, 0xC0, 0x3B, 0x00,
    WRITE_C8_D16, 0xC1, 0x10, 0x0C,
    WRITE_C8_D16, 0xC2, 0x07, 0x0A,
    WRITE_C8_D8, 0xC7, 0x00,
    WRITE_C8_D8, 0xC7, 0x10,
    WRITE_C8_D8, 0xCD, 0x08,

    WRITE_COMMAND_8, 0xB0, // Positive Voltage Gamma Control
    WRITE_BYTES, 16,
    0x05, 0x12, 0x98, 0x0e, 0x0F,
    0x07, 0x07, 0x09, 0x09, 0x23,
    0x05, 0x52, 0x0F, 0x67, 0x2C,
    0x11,



    WRITE_COMMAND_8, 0xB1, // Negative Voltage Gamma Control
    WRITE_BYTES, 16,
    0x0B, 0x11, 0x97, 0x0C, 0x12,
    0x06, 0x06, 0x08, 0x08, 0x22,
    0x03, 0x51, 0x11, 0x66, 0x2B,
    0x0F,

    WRITE_COMMAND_8, 0xFF,
    WRITE_BYTES, 5, 0x77, 0x01, 0x00, 0x00, 0x11,

    WRITE_C8_D8, 0xB0, 0x5D,
    WRITE_C8_D8, 0xB1, 0x2D,
    WRITE_C8_D8, 0xB2, 0x81,
    WRITE_C8_D8, 0xB3, 0x80,
    WRITE_C8_D8, 0xB5, 0x4E,
    WRITE_C8_D8, 0xB7, 0x85,
    WRITE_C8_D8, 0xB8, 0x20,
    WRITE_C8_D8, 0xC1, 0x78,
    WRITE_C8_D8, 0xC2, 0x78,

    WRITE_C8_D8, 0xD0, 0x88,

    WRITE_COMMAND_8, 0xE0,
    WRITE_BYTES, 4,
    0x00, 0x00, 0x02,

    WRITE_COMMAND_8, 0xE1,
    WRITE_BYTES, 16,
    0x06, 0x30, 0x08, 0x30, 0x05,
    0x30, 0x07, 0x30, 0x00, 0x33,
    0x33,

    WRITE_COMMAND_8, 0xE2,
    WRITE_BYTES, 16,
    0x11, 0x11, 0x33, 0x33, 0xf4,
    0x00, 0x00, 0x00, 0xf4, 0x00,
    0x00, 0x00,

    WRITE_COMMAND_8, 0xE3,
    WRITE_BYTES, 4, 0x00, 0x00, 0x11, 0x11,

    WRITE_C8_D16, 0xE4, 0x44, 0x44,

    WRITE_COMMAND_8, 0xE5,
    WRITE_BYTES, 16,
    0x0d, 0xf5, 0x30, 0xf0, 0x0f,
    0xf7, 0x30, 0xf0, 0x09, 0xf1,
    0x30, 0xf0, 0x0b, 0xf3, 0x30, 0xf0,

    WRITE_COMMAND_8, 0xE6,
    WRITE_BYTES, 4, 0x00, 0x00, 0x11, 0x11,

    WRITE_C8_D16, 0xE7, 0x44, 0x44,

    WRITE_COMMAND_8, 0xE8,
    WRITE_BYTES, 16,
    0x0c, 0xf4, 0x30, 0xf0,
    0x0e, 0xf6, 0x30, 0xf0,
    0x08, 0xf0, 0x30, 0xf0,
    0x0a, 0xf2, 0x30, 0xf0,

    WRITE_C8_D8, 0xe9, 0x36,

    WRITE_COMMAND_8, 0xEB,
    WRITE_BYTES, 7,
    0x00, 0x01, 0xe4, 0xe4,
    0x44, 0x88, 0x40,

    WRITE_COMMAND_8, 0xED,
    WRITE_BYTES, 16,
    0xff, 0x10, 0xaf, 0x76,
    0x54, 0x2b, 0xcf, 0xff,
    0xff, 0xfc, 0xb2, 0x45,
    0x67, 0xfa, 0x01, 0xff,

    WRITE_COMMAND_8, 0xEF,
    WRITE_BYTES, 6,
    0x08, 0x08, 0x08, 0x45,
    0x3f, 0x54,

    WRITE_COMMAND_8, 0xFF,
    WRITE_BYTES, 5, 0x77, 0x01, 0x00, 0x00, 0x00,

    WRITE_COMMAND_8, 0x11, // Sleep Out
    END_WRITE,

    DELAY, 150,

    BEGIN_WRITE,
    WRITE_C8_D8, 0x3A, 0x66,
    WRITE_C8_D8, 0x36, 0x08,
    WRITE_C8_D8, 0x35, 0x00,
    WRITE_COMMAND_8, 0x29, // Display On
    END_WRITE,
    DELAY, 20
};
`

**Reply by T-Bone-one** [reply]
```
`#include` <Arduino_GFX_Library.h>

Arduino_DataBus *bus = new Arduino_SWSPI(GFX_NOT_DEFINED, 42, 2, 1, GFX_NOT_DEFINED);

Arduino_ESP32RGBPanel *rgbpanel = new Arduino_ESP32RGBPanel(
  40 /* DE */, 39 /* VSYNC */, 38 /* HSYNC */, 41 /* PCLK */,
  5 /* R0 */, 45 /* R1 */, 48 /* R2 */, 47 /* R3 */, 21 /* R4 */,
  14 /* G0 */, 13 /* G1 */, 12 /* G2 */, 11 /* G3 */, 10 /* G4 */, 9 /* G5 */,
  46 /* B0 */, 3 /* B1 */, 8 /* B2 */, 18 /* B3 */, 17 /* B4 */,
  1 /* hsync_polarity */, 22 /* hsync_front_porch */, 24 /* hsync_pulse_width */, 6 /* hsync_back_porch */,
  1 /* vsync_polarity */, 16 /* vsync_front_porch */, 4 /* vsync_pulse_width */, 10 /* vsync_back_porch */);


Arduino_RGB_Display *gfx = new Arduino_RGB_Display(
  480 /* width */, 480 /* height */, rgbpanel, 0 /* rotation */, true /* auto_flush */,
  bus, GFX_NOT_DEFINED /* RST */, st7701_type6_init_operations, sizeof(st7701_type6_init_operations));


void setup(void) {
  Serial.begin(115200);

  Serial.println("Arduino_GFX Hello World example");
  gfx->invertDisplay(2);
#ifdef GFX_EXTRA_PRE_INIT
  GFX_EXTRA_PRE_INIT();
#endif

  if (!gfx->begin()) {
    Serial.println("gfx->begin() failed!");
  }
  gfx->fillScreen(BLACK);

#ifdef GFX_BL
  pinMode(GFX_BL, OUTPUT);
  digitalWrite(GFX_BL, HIGH);
#endif

  gfx->setCursor(150, 150);
  gfx->setTextSize(2);
  gfx->setTextColor(RED);
  gfx->println("Hello World! in RED");
  gfx->setCursor(150, 180);
  gfx->setTextColor(BLUE);
  gfx->println("Hello World! in BLUE");
  gfx->setCursor(150, 210);
  gfx->setTextColor(GREEN);
  gfx->println("Hello World! in GREEN");
  gfx->setCursor(150, 240);
  gfx->setTextColor(YELLOW);
  gfx->println("Hello World! in YELLOW");
  delay(5000);  // 5 seconds
    gfx->fillScreen(BLACK);
}

void loop() {
  gfx->setCursor(random(gfx->width()), random(gfx->height()));
  gfx->setTextColor(random(0xffff), random(0xffff));
  gfx->setTextSize(random(6) /* x scale */, random(6) /* y scale */, random(2) /* pixel_margin */);
  gfx->println("Hello World!");

  delay(1000); // 1 second
}`
```

**Reply by T-Bone-one** [reply]
This works for me as it should.

**Reply by jfseaman** [reply]
Thank you.

**Comment by showdown** [comment]
Hi, I'm trying to get a round 2.8 screen up and running with the ESP LCD board.  My screen and board are identical to yours but the code doesn't compile.  I get this error: 

error: 'DE' was not declared in this scope

Any thoughts?

**Reply by T-Bone-one** [reply]
[HelloWorldroundworking.zip](https://github.com/user-attachments/files/17144623/HelloWorldroundworking.zip)

Code is not formated correctly .. try file in attachement.

**Reply by showdown** [reply]
Awesome, code compiles perfectly, thanks!!

Having trouble connecting to the board itself- I keep getting fatal errors.  This is my first ESP board and I'm reading that I have to push the boot or reset button during upload...?  Any tips on your setup, what settings did you have in IDE to get it to upload?

**Reply by T-Bone-one** [reply]
use Esp Boardmanager V2.0.17 
push and hold BOOT while pushing (short) RESET ... then upload Code 
Board settings in IDE :
USB CDC on Boot : Enabled
Flashsize: 8MB (64MB)
PSRAM : OPIPSRAM
